src/cli/Hibp.cpp

```diff
@@ -0,0 +1,221 @@
+
+#include "Hibp.h"
+#include "cli/Utils.h"
+
+#include <QCommandLineParser>
+#include <QFile>
+
+#include "cli/TextStream.h"
+#include "core/Database.h"
+#include "core/Entry.h"
+#include "core/Group.h"
+#include "core/Tools.h"
+#include "crypto/CryptoHash.h"
+
+namespace
+{
+
+    struct Sha1Hash
+    {
+        uchar m_data[20];
+
+        Sha1Hash()
+        {
+        }
+
+        Sha1Hash(const QByteArray& a)
+        {
+            Q_ASSERT(a.size() == sizeof(m_data));
+            std::memcpy(&m_data[0], a.data(), sizeof(m_data));
+        }
+
+        bool operator==(const Sha1Hash& other) const
+        {
+            return std::memcmp(&m_data[0], &other.m_data[0], sizeof(m_data)) == 0;
+        }
+    };
+
+    uint qHash(const Sha1Hash& b)
+    {
+        uint result;
+        std::memcpy(&result, &b.m_data[0], sizeof(result));
+        return result;
+    }
+
+    bool parseHibpLine(QFile& file, Sha1Hash& sha1, size_t& numPwns, bool& atEnd, QString& error)
+    {
+        auto failedTo = [&](const QString& action) {
+            error = QObject::tr("failed to %1").arg(action);
+            return false;
+        };
+
+        auto failedToIo = [&](const QString& action) {
+            error = QObject::tr("failed to %1: %2").arg(action).arg(file.errorString());
+            return false;
+        };
+
+        atEnd = false;
+
+        char hexSha1[sizeof(Sha1Hash::m_data) * 2];
+        qint64 rc = file.read(&hexSha1[0], sizeof(hexSha1));
+        if (rc == -1) {
+            return failedToIo(QObject::tr("read SHA-1"));
+        } else if (rc == 0) {
+            atEnd = true;
+            return true;
+        } else if (rc != sizeof(hexSha1)) {
+            return failedTo(QObject::tr("read SHA-1 (short read)"));
+        }
+
+        if (!Tools::hexDecode(&sha1.m_data[0], sizeof(sha1.m_data), &hexSha1[0], sizeof(hexSha1))) {
+            return failedTo(QObject::tr("hex-decode SHA-1"));
+        }
+
+        char c;
+        if (!file.getChar(&c)) {
+            return failedToIo(QObject::tr("read ':' character"));
+        }
+        if (c != ':') {
+            return failedTo(QObject::tr("parse line (expected ':')"));
+        }
+
+        numPwns = 0;
+        while (true) {
+            if (!file.getChar(&c)) {
+                return failedToIo(QObject::tr("read pwnage count"));
+            }
+
+            if (c == '\n' || c == '\r') {
+                break;
+            }
+
+            if (!('0' <= c && c <= '9')) {
+                return failedTo(QObject::tr("parse pwnage count"));
+            }
+
+            numPwns *= 10;
+            numPwns += (c - '0');
+        }
+
+        while (1 == file.peek(&c, 1) && (c == '\n' || c == '\r')) {
+            file.getChar(&c);
+        }
+
+        return true;
+    }
+
+    QString pwnageReport(const Entry& entry, size_t numPwns)
+    {
+        QStringList ancestors;
+        for (auto g = entry.group(); g && g != g->database()->rootGroup(); g = g->parentGroup()) {
```

Comment:
A while loop might be easier to understand here. You should also assemble the path string directly. No need to save components to a `QStringList` first and then iterate that again.

Reply:
The `QStringList` variable is necessary because I need the ancestor path in the reverse order to which I obtain it in this loop.

Reply:
QString has a prepend method. But I guess I'm be fine with the string list approach, too. It may save a few extra copy operations at the cost of more allocations.